tests/baselines/reference/generatorYieldContextualType.types

```diff
@@ -25,11 +25,11 @@ declare function f2<T, R, S>(gen: () => Generator<R, T, S> | AsyncGenerator<R, T
 f2<0, 0, 1>(async function* () {
 >f2<0, 0, 1>(async function* () {	const a = yield 0;	return 0;}) : void
 >f2 : <T, R, S>(gen: () => Generator<R, T, S> | AsyncGenerator<R, T, S>) => void
->async function* () {	const a = yield 0;	return 0;} : () => AsyncGenerator<0, 0, 1>
+>async function* () {	const a = yield 0;	return 0;} : () => AsyncGenerator<0, 0, 1 | undefined>
 
 	const a = yield 0;
->a : 1
->yield 0 : 1
+>a : 1 | undefined

```

Comment:
In a sense, this is not wrong—`gen` could have its `next()` called with no arguments, making `a` `undefined` at runtime. But previously, this fell into a special case of ignoring that optionality for convenience.

Now, when we try to get the iteration types of `Generator<R, T, S> | AsyncGenerator<R, T, S>` as an async _iterable_, we get `noIterationTypes`, because the former doesn’t have a `[Symbol.asyncIterator]()` member. Instead, we fall back to getting the iteration types of this union as an async _iterator_, using its next/return/throw method signatures to determine `T`, `TReturn`, and `TNext`. Because we have a union signature, it doesn’t count as a reference to one of the global generator/iterator types, and so it doesn’t take the fast path that ignores the optionality. Instead, it correctly sees that the `next()` method can take no arguments or `[1]`, and so it infers the type `1 | undefined`.

More work could be done to make union signatures take the fast path that drops the optionality if every signature in the union is a reference to one of the global generator/iterator types, but I felt like this was getting fairly complex for limited utility—the results it produces with less complexity are strictly more correct given the limitations we have for modeling this.